Sources/PenguinStructures/Tuple.swift

```diff
@@ -0,0 +1,144 @@
+//******************************************************************************
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+/// Generalized algebraic product types
+///
+/// Swift's built-in tuple types are algebraic product types, but since they are
+/// not nominal and not easily decomposed, they don't lend themselves to many
+/// types of useful processing.  Models of `Tuple` don't have those problems.
+public protocol AlgebraicProduct {
+  /// The type of the first element.
+  associatedtype Head
+
+  /// An algebriac product formed by composing the remaining elements.
+  associatedtype Tail: AlgebraicProduct
+
+  /// The first element.
+  var head: Head { get set }
+  
+  /// All elements but the first.
+  var tail: Tail { get set }
+}
+
+extension Empty: AlgebraicProduct {
+  /// The first element, when `self` is viewed as an instance of algebraic
+  /// product type.
+  public var head: Never { get {  fatalError() } set {  } }
```

Comment:
Should `set` also `fatalError()`?

Reply:
Neither set nor get can be called; the `fatalError` is just there to make the compiler happy.